src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs

```diff
@@ -2,381 +2,313 @@
+            return 0;
         }
 
-        // Helper to get high 32-bit of 64-bit int
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static uint Hi32Bits(long a)
+        [RuntimeExport("Dbl2LngOvf")]
+        public static long Dbl2LngOvf(double value)
         {
-            return (uint)(a >> 32);
+            const double two63 = IntMaxValueOffset * UIntMaxValueOffset;
+
+            // Note that this expression also works properly for val = NaN case
+            // We need to compare with the very next double to two63. 0x402 is epsilon to get us there.
+            if (value is > -two63 - 0x402 and < two63)
+            {
+                return (long)value;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
         }
 
-        // Helper to get high 32-bit of 64-bit int
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static uint Hi32Bits(ulong a)
+        [RuntimeExport("Dbl2ULngOvf")]
+        public static ulong Dbl2ULngOvf(double value)
         {
-            return (uint)(a >> 32);
+            const double two64 = UIntMaxValueOffset * UIntMaxValueOffset;
+            // Note that this expression also works properly for val = NaN case
+            if (value is > -1.0 and < two64)
+            {
+                return (ulong)value;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
         }
 
+#if !TARGET_64BIT
+        //
+        // 64-bit checked multiplication for 32-bit platforms
+        //
+
+        private const string RuntimeLibrary = "*";
+
         [RuntimeExport("LMulOvf")]
-        public static long LMulOvf(long i, long j)
+        public static long LMulOvf(long left, long right)
         {
-            long ret;
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static uint High32Bits(ulong a)
+            {
+                return (uint)(a >> 32);
+            }
+
+#if DEBUG
+            long result = left * right;
+#endif
 
             // Remember the sign of the result
-            int sign = (int)(Hi32Bits(i) ^ Hi32Bits(j));
+            int sign = (int)(High32Bits((ulong)left) ^ High32Bits((ulong)right));
 
             // Convert to unsigned multiplication
-            if (i < 0) i = -i;
-            if (j < 0) j = -j;
+            if (left < 0)
+                left = -left;
+            if (right < 0)
+                right = -right;
 
             // Get the upper 32 bits of the numbers
-            uint val1High = Hi32Bits(i);
-            uint val2High = Hi32Bits(j);
+            uint val1High = High32Bits((ulong)left);
+            uint val2High = High32Bits((ulong)right);
 
             ulong valMid;
 
             if (val1High == 0)
             {
                 // Compute the 'middle' bits of the long multiplication
-                valMid = Mul32x32To64(val2High, (uint)i);
+                valMid = (ulong)val2High * (uint)left;
```

Comment:
I think the code was easier to understand with the `Mul32x32To64` helper function. I do not think it is an improvement to inline it like this.